server/container_execsync.go

```diff
@@ -17,12 +16,7 @@ func (s *Server) ExecSync(ctx context.Context, req *pb.ExecSyncRequest) (*pb.Exe
 		return nil, status.Errorf(codes.NotFound, "could not find container %q: %v", req.ContainerId, err)
 	}
 
-	if err := s.Runtime().UpdateContainerStatus(c); err != nil {
-		return nil, err
-	}
-
-	cState := c.State()
-	if !(cState.Status == oci.ContainerStateRunning || cState.Status == oci.ContainerStateCreated) {
+	if !c.IsAlive() {
```

Comment:
maybe

```go
if _, err := c.verifyPid(); err != nil {
```

and remove the whole `IsAlive` function?

Or, use `IsAlive` in all the places where you're using the `if` like the one in this comment. Otherwise it's a duplicated functionality.

Reply:
the nice thing about the different functions has to do with error handling. In the case of IsAlive, I believe it should be used in cases we presume the container to be alive, and it would be an error if it was not.

Conversely, verifyPid does not make an assertion of what the liveliness of the container should be

For instance, all uses in runtime_oci.go assume the container will eventually not be alive. in that case, we wouldn't want to print an error that the container wasn't alive

I think it is correct to error in `server/container_execsync.go`, as the client asked to exec in a container that is not running. Propagating the reason for this failure (pid wrap, container exited, we exec'ed too early somehow) is useful.

Do you have suggestions on how to balance these two use cases and error scenerios @kolyshkin